---
  html_document:
    toc: true
    highlight: zenburn
    theme: united
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview
This is looking at some RNA-seq data from Andrew Lassar. The RNA comes from
the artcular cartilage of the cow. There are four samples, one from
the superior zone, one from the deep zone, one from the superior zone
and treated with BzATP and forskolin and one from the deep zone
treated with BzATP and forskolin. There is some evidence the treatment was not
effective as a marker gene for the treatment is missing.

There are no replicates in this data or the treatment and zone interaction,
so we can't really do a differential analysis for that very well. Andrew
indicated he was interested in looking at the difference between the
two zones, and we can do that if we regress out the treatment effect. This
will give us a replicate for the tissue effect that way.

We processed the reads using the cow genome, build UMD3.1 using the
version 83 of the Ensembl annotation.

We ran these samples through bcbio-nextgen 0.9.7a, using the STAR aligner to
align the reads. The samples were quantitated using the counts estimated by
Sailfish at the level of the transcript, and then summed the counts per
transcript to generate counts per gene.

```{r check-library-functions, echo=FALSE}
check_install = function(packages) {
   not_installed = setdiff(packages, rownames(installed.packages()))
   if(length(not_installed) > 0) {
      write(paste("The libraries", not_installed, "are not installed, aborting.",
                   sep=" "), stdout())
      stop()
   }
}
```

```{r load-libraries, echo=FALSE}
packages = c("ggplot2", "reshape", "gplots", "edgeR", "CHBUtils", "pheatmap",
              "DESeq2", "tximport", "logging", "DT")
check_install(packages)
installed = lapply(packages, library, character.only = TRUE)
```

```{r qc-setup}
basicConfig()
project_summary = "/Users/rory/cache/lassar-bovine-rnaseq/results/2016-02-19_lassar-bovine/project-summary.csv"
counts_file = "/Users/rory/cache/lassar-bovine-rnaseq/results/2016-02-19_lassar-bovine/combined.counts"
tx2genes_file = "/Users/rory/cache/lassar-bovine-rnaseq/results/2016-02-19_lassar-bovine/tx2gene.csv"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
if (file.exists(tx2genes_file)) {
  loginfo("Using gene counts calculated from the Sailfish transcript counts.")
  sf_files = file.path("..", "..", rownames(summarydata), "sailfish",
                      rownames(summarydata), "quant.sf")
  names(sf_files) = rownames(summarydata)
  tx2gene = read.table(tx2genes_file, sep=",", row.names=NULL, header=FALSE)
  txi.salmon = tximport(sf_files, type="salmon", tx2gene=tx2gene,
                        reader=readr::read_tsv, countsFromAbundance="lengthScaledTPM")
  counts = round(data.frame(txi.salmon$counts, check.names=FALSE))
  write.table(counts, file="sailfish_counts.tsv", col.names=TRUE, row.names=TRUE,
              quote=FALSE, sep="\t")
} else {
  loginfo("Using gene counts calculated from featureCounts.")
  counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
}
counts = counts[, order(colnames(counts)), drop=FALSE]
colnames(counts) = gsub(".counts", "", colnames(counts))

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Median.insert.size", "Mapped.reads.pct", "Total.reads",
    "avg_coverage_per_region", "Mapped.Reads")
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
```

```{r sanitize-datatable}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}
```

```{r set-random-seed}
# set seed for reproducibility
set.seed(1454944673)
```

```{r conversions}
library(biomaRt)
bovine = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                dataset="btaurus_gene_ensembl",
                host = "jul2015.archive.ensembl.org")
conversions = getBM(attributes=c("ensembl_gene_id", "hgnc_symbol", "gene_biotype",
                                 "external_gene_name", "description"), mart=bovine)
```

## Sample metadata
This table contains a bunch of columns we estimate from the STAR alignments
about each sample, along with the provided metadata about the identity of the sample.

```{r summary-table, echo=FALSE, cache=FALSE}
sanitize_datatable(summarydata, style='bootstrap')
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

```{r qualimap-run}
qualimap_run = "Mapped" %in% colnames(summarydata)
```

## Mapped reads
This is a good amount of mapped reads per sample, around 20-40 million reads per
sample. It does not look like any of the samples failed. This amount of variation
in the amount of reads per sample is normal. This is a good sign that the
quality of the libraries was high.

```{r mapped-plot, eval=qualimap_run}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

```{r mapped-plot-noqualimap, eval=!qualimap_run}
ggplot(summarydata, aes(x=Name, y=Mapped.reads)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

## Genomic mapping rate
The genomic mapping rate is excellent, another good sign.

```{r mapping-rate-plot, eval=qualimap_run}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

```{r mapping-rate-plot-noqualimap, eval=!qualimap_run}
ggplot(summarydata, aes(x=Name, y=Mapped.reads.pct)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
The number of genes detected is high too. There are `r nrow(conversions)`
annotated genes in version 83 of the Ensembl annotation of the cow, and
about half of those are detected in these samples. That is a good sign.

```{r genes-detected-plot}
dd = data.frame(Name=colnames(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Exonic mapping rate
The exonic mapping rate is low, for RNA-seq we'd expect this to be higher,
in the 70% range or higher. Usually this would be a sign of some kind of
DNA contamination in the sample. I think in this instance it just indicates
that the cow genome has an incomplete transcript annotation.

```{r exonic-mapping-plot, eval="Exonic.Rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
rRNA are nuisance genes in RNA-seq experiments and they should be removed in
most library preparation kits. We rely on the biotype annotations from Ensembl
to identify these genes. Here we can see the amount of rRNA contamination that
we can identify is low which is good.

```{r check-rRNA-eval}
eval_rRNA = "rRNA_rate" %in% colnames(summarydata) & !sum(is.na(summarydata$rRNA_rate)) == nrow(summarydata)
```
```{r rRNA-rate-plot, eval=eval_rRNA}
ggplot(summarydata, aes(x=Name, y=rRNA_rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Boxplot of log10 counts per gene
The samples have pretty similar count distributions, which good.

```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

These samples started out looking pretty reasonable, and normalizing keeps them
looking reasonable.

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
Here we can see the deep untreated sample has much less genes with a very
low level of expression. We could see it in the boxplots above too.
This is probably not biological, and more a library preparation difference.
This is a good example of why we have to have replicates to look at differences
of the treatment in the two zones.

```{r density-normalized}
ggplot(melted, aes(x=count, group=sample, color=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## PCA plot
This is a PCA plot of the samples, we first smooth out the mean-variance
relationship that exists in RNA-seq and then find the most variable genes. Then
we use those variable genes to do PCA. PCA reduces the data down to a set of
components that are orthogonal to each other and describe the largest amount of
the variance between the most variant genes. The first component describes the
most of the variance, the second component the second most, and so on.

Below we can see the first component separates the samples out based on
the zone. The second componeny separates the samples out based on treatment.
So it does look like there is both a zone and a treatment effect.

Having a small number of replicates means it is hard to draw strong conclusions
from the plot, though-- the four points have to be somewhere on the plot, and
can easily end up in this configuration by chance. However it is a good sign.

```{r pca, eval=ncol(counts) > 1}
dds = DESeqDataSetFromMatrix(countData=counts, colData=summarydata, design=~Name)
vst = varianceStabilizingTransformation(dds)
plotPCA(vst, intgroup=c("Name"))
```

```{r deseq2-tximport-helper-functions}
# snagged from development version of DESeq
DESeqDataSetFromTximport <- function(txi, colData, design, ...)
{
  counts <- round(txi$counts)
  mode(counts) <- "integer"
  dds <- DESeqDataSetFromMatrix(countData=counts, colData=colData, design=design, ...)
  stopifnot(txi$countsFromAbundance %in% c("no","scaledTPM","lengthScaledTPM"))
  if (txi$countsFromAbundance %in% c("scaledTPM","lengthScaledTPM")) {
    message("using length scaled TPM counts from tximport")
  } else {
    message("using counts and average transcript lengths from tximport")
    lengths <- txi$length
    dimnames(lengths) <- dimnames(dds)
    assays(dds)[["avgTxLength"]] <- lengths
  }
  return(dds)
}

subset_tximport = function(txi, rows, columns) {
    txi$counts = txi$counts[rows, columns]
    txi$abundance = txi$abundance[rows, columns]
    txi$length = txi$length[rows, columns]
    return(txi)
}
```


## Summary wrap-up
The samples look like high quality RNA-seq samples and they separate how we
expected them to in the PCA plot.

# Differential expression
Here do a differential expression analysis using DESeq2. We set up a model
that looks like `~treated+zone` to fit to the gene expression values. Set
up this way, we can tease out the part of the gene expression differences
that are due to the treatment by looking at the coefficients of `treatment`,
and the gene expression differences due to the zone the sample is from by
looking at the coefficients of `zone`.

```{r de-setup}
library(DEGreport)
library(vsn)
design = ~treated+zone
condition = "zone"
```
```{r deseq2-expression-analysis, results='asis'}
counts <- counts[rowSums(counts>0)>1,]
if(exists("txi.salmon")) {
   loginfo("Using Sailfish gene counts for the DESeq2 model.")
   txi.salmon = subset_tximport(txi.salmon, rownames(counts), colnames(counts))
   dds = DESeqDataSetFromTximport(txi.salmon, colData=summarydata, design=design)
} else {
  loginfo("Using counts from featureCounts for the DESeq2 model.")
  dds = DESeqDataSetFromMatrix(countData=counts,
      colData=summarydata, design = design  )
}
geoMeans = apply(counts, 1, function(row) if (all(row == 0)) 0 else
                 exp(mean(log(row[row != 0]))))
dds = estimateSizeFactors(dds, geoMeans=geoMeans)
dds = DESeq(dds)
```

## Effect of variance stabilization

```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1))
meanSdPlot(assay(rld[notAllZero,]))
meanSdPlot(assay(vsd[notAllZero,]))
```

## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds)
```

```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column) {
  all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse=" vs ")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```

### Effect of zone
Below we look at the effect of zone on the expression of genes in the
cartilage. Positive fold changes are
genes higher in the deep layers compared to the superficial layers.
Negative fold changes are genes lower in the deep layers compared
to the superficial layers.

#### MA-plots (zone)
This plots the log fold change vs the mean expression. We can see quite a few
genes are called differentially expressed in the superficial vs. deep zone.

```{r DESeq-output, results='asis'}
all_results = handle_deseq2(dds, summarydata, condition)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

#### Volcano-plots (zone)
Another way to display the hits, here we look at the log2 fold change on the
x-axis vs the p-value on the y-axis.

```{r DESeq-volcano}
for(i in seq(length(all_results))) {
  stats = as.data.frame(all_results[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=names(all_results)[i], lfc.cutoff=1.5)
  print(p)
}
```

#### DEGreport (zone)

```{r get-groups, echo=FALSE}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp," "))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[3]]
  list(g1,g2)
}
```

#### Pvalues-vs-Mean (zone)

Here we plot some information about how the p-values are correlated with the
mean or the standard deviation. We'd expect there to be an even distribution
in quartiles for each p-value cutoff, showing there isn't a strong bias
towards finding genes with higher mean expression.

```{r DEGreport-M}
plots = list()
scale_factor = round(1/nr * 14)
for(i in seq(length(all_results))) {
  plots[[i]] = degMean(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Mean for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

#### Pvalues-vs-Variation (zone)
Similar to the plot above, but making sure we are not calling differential
expression that isn't biased towards calling differential expression of only
highly variable genes.

```{r DEGreport-V}
plots = list()
for(i in seq(length(all_results))) {
  plots[[i]] = degVar(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

#### Differentially expressed genes (zone)
Here we spit out the genes called differentially expressed between the
two zones. Genes with a positive log fold change are higher in the deep samples
compared to the superficial samples.

```{r DESeq-tables, echo=FALSE, cache=FALSE}
library(dplyr)
zone_res = as.data.frame(all_results[[1]])
zone_res$id = rownames(zone_res)
zone_res = zone_res %>% left_join(conversions, by=c("id"="ensembl_gene_id"))
write.table(zone_res, file=paste(names(all_results)[i], ".tsv", sep=""),
                        quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
zone_sig = subset(zone_res, padj < 0.05)
sanitize_datatable(zone_sig)
```

There are `r nrow(zone_sig)` differentially expressed between the two
zones, using an adjusted p-value cutoff of 0.05. `r nrow(subset(zone_sig, log2FoldChange > 0))` are more highly expressed in the superficial layer and
`r nrow(subset(zone_sig, log2FoldChange < 0))` are more highly expressed in the
deep layer.

I spit out a file called [deep vs superficial.tsv](deep vs superficial.tsv) which
has this information in a table format.

### Effect of treatment
Below we look at the effect of zone on the expression of genes in the
cartilage. Positive fold changes are
genes higher in the superficial layers compared to the deep layers.
Negative fold changes are genes lower in the superficial layers compared
to the deep layers.

#### MA-plots (treatment)
This plots the log fold change vs the mean expression.

```{r DESeq-output-treatment, results='asis'}
condition = "treated"
all_results = handle_deseq2(dds, summarydata, condition)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

#### Volcano-plots (treatment)
Another way to display the hits, here we look at the log2 fold change on the
x-axis vs the p-value on the y-axis.

```{r DESeq-volcano-treatment}
for(i in seq(length(all_results))) {
  stats = as.data.frame(all_results[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=names(all_results)[i], lfc.cutoff=1.5)
  print(p)
}
```

#### Pvalues-vs-Mean (treatment)

Here we plot some information about how the p-values are correlated with the
mean or the standard deviation. We'd expect there to be an even distribution
in quartiles for each p-value cutoff, showing there isn't a strong bias
towards finding genes with higher mean expression.

```{r DEGreport-M-treatment}
plots = list()
scale_factor = round(1/nr * 14)
for(i in seq(length(all_results))) {
  plots[[i]] = degMean(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Mean for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

#### Pvalues-vs-Variation (treatment)
Similar to the plot above, but making sure we are not calling differential
expression that isn't biased towards calling differential expression of only
highly variable genes.

```{r DEGreport-V-treatment}
plots = list()
for(i in seq(length(all_results))) {
  plots[[i]] = degVar(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

#### Differentially expressed genes (treatment)
Here we spit out the genes called differentially expressed between the
two treatments. Genes with a positive fold change are higher in the treated
samples.

```{r DESeq-tables-treatment, echo=FALSE, cache=FALSE}
library(dplyr)
treatment_res = as.data.frame(all_results[[1]])
treatment_res$id = rownames(treatment_res)
treatment_res = treatment_res %>% left_join(conversions, by=c("id"="ensembl_gene_id"))
write.table(treatment_res, file=paste(names(all_results)[1], ".tsv", sep=""),
                        quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
treatment_sig = subset(treatment_res, padj < 0.05)
sanitize_datatable(treatment_sig)
```

There are `r nrow(treatment_sig)` differentially expressed between the two
treatments, using an adjusted p-value cutoff of 0.05. `r nrow(subset(treatment_sig, log2FoldChange > 0))` are more highly expressed in the treated samples and
`r nrow(subset(treatment_sig, log2FoldChange < 0))` are more highly expressed in the
untreated samples.

I spit out a file called [treated vs untreated.tsv](treated vs untreated.tsv)
which has this information in a table format.

### Wrap-up
While we can call differences between the treated/untreated and the
superficial/deep zones, these could very easily be spurious results. We don't
have a great way of determining how robust the finding is with only a couple
of replicates. Furthermore, with no replicates our power is very low, so we
are most likely missing things that are really there, we just don't have
the power to see them.

Let me know if these results make any sense to you in a biological context.
Do you see any hits that you were expecting to see? I don't know anything
about the biology of your experiment so you'll have to make those judgements.

#### Download zone results
[deep vs superficial.tsv](deep vs superficial.tsv)

#### Download zreatment results
[treated vs untreated.tsv](treated vs untreated.tsv)

#### Count matrix
[sailfish counts](sailfish_counts.tsv)
